quickwit/quickwit-serve/src/delete_task_api/handler.rs

```diff
@@ -0,0 +1,196 @@
+    pub query: String,
+    // Fields to search on
+    #[serde(default)]
+    pub search_fields: Vec<String>,
+    /// If set, restrict delete to documents with a `timestamp >= start_timestamp`.
+    pub start_timestamp: Option<i64>,
+    /// If set, restrict delete to documents with a `timestamp < end_timestamp``.
+    pub end_timestamp: Option<i64>,
+}
+
+/// Delete query API handlers.
+pub fn delete_task_api_handlers(
+    metastore: Arc<dyn Metastore>,
+    delete_task_service_mailbox_opt: Option<Mailbox<DeleteTaskService>>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    get_delete_tasks_handler(metastore.clone(), delete_task_service_mailbox_opt.clone()).or(
+        post_delete_tasks_handler(metastore, delete_task_service_mailbox_opt),
+    )
+}
+
+pub fn get_delete_tasks_handler(
+    metastore: Arc<dyn Metastore>,
+    delete_task_service_mailbox: Option<Mailbox<DeleteTaskService>>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    warp::path!(String / "delete-tasks")
+        .and(warp::get())
+        .and(with_arg(metastore))
+        .and(require(delete_task_service_mailbox))
+        .and_then(get_delete_tasks)
+}
+
+// Returns delete tasks in json format for a given `index_id`.
+// Note that `_delete_task_service_mailbox` is not used...
+// Explanation: we don't want to expose any delete tasks endpoints without a running
+// `DeleteTaskService`. This is ensured by requiring a `Mailbox<DeleteTaskService>` in
+// `get_delete_tasks_handler` and consequently we get the mailbox in `get_delete_tasks` signature.
+async fn get_delete_tasks(
+    index_id: String,
+    metastore: Arc<dyn Metastore>,
+    _delete_task_service_mailbox: Mailbox<DeleteTaskService>,
+) -> Result<impl warp::Reply, Infallible> {
+    let delete_tasks = metastore.list_delete_tasks(&index_id, 0).await;
+    Ok(Format::PrettyJson.make_rest_reply(delete_tasks))
+}
+
+pub fn post_delete_tasks_handler(
+    metastore: Arc<dyn Metastore>,
+    delete_task_service_mailbox: Option<Mailbox<DeleteTaskService>>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    warp::path!(String / "delete-tasks")
+        .and(warp::body::json())
+        .and(warp::post())
+        .and(with_arg(metastore))
+        .and(require(delete_task_service_mailbox))
+        .and_then(post_delete_request)
+}
+
+async fn post_delete_request(
+    index_id: String,
+    delete_request: DeleteQueryRequest,
+    metastore: Arc<dyn Metastore>,
+    delete_task_service_mailbox: Mailbox<DeleteTaskService>,
+) -> Result<impl warp::Reply, Infallible> {
+    let delete_query = DeleteQuery {
+        index_id: index_id.clone(),
+        start_timestamp: delete_request.start_timestamp,
+        end_timestamp: delete_request.end_timestamp,
+        query: delete_request.query,
+        search_fields: delete_request.search_fields,
+    };
+    let delete_task = metastore.create_delete_task(delete_query).await;
```

Comment:
I don't understand why you edit add the delete task in the handler.
I expected that to be done in the delete task service.

Right now trying to put a sentence on what `NewDeleteTask` does is really clumsy.

Reply:
Indeed, I hesitated on this, my rationale here was "I generally prefer not hiding simple metastore operations". And creating a task is super simple so I found it nice to call it directly.
But your point is very relevant... `NewTask` name is bad. A side effect was also I need to add a comment to explain the require...
So I think I will move this operation inside the delete service.